### ISSUE TITLE
specifying fields in get_leads is broken

### ISSUE THREAD

**Issue by thomasdziedzic-pd** [raszi/mrkt#48]
Setup:
ruby 2.4.1
mrkt 0.10.0

Problem:
When I specify fields when calling get_leads, the response doesn't include the asked for fields.

```ruby
pry(main)> response = $marketo.get_leads(:email, ['REDACTED'], fields: ['phone'])
GET https://REDACTED.mktorest.com/rest/v1/leads.json?fields%5B%5D=phone&filterType=email&filterValues=REDACTED
User-Agent: "Faraday v0.15.4"
Authorization: "Bearer REDACTED"
Status 200
server: "nginx"
date: "Tue, 16 Jul 2019 23:17:18 GMT"
content-type: "application/json;charset=UTF-8"
content-length: "210"
connection: "close"
{:requestId=>"REDACTED",
 :result=>
  [{:id=>REDACTED,
    :firstName=>"REDACTED",
    :lastName=>"REDACTED",
    :email=>"REDACTED",
    :updatedAt=>"2019-06-19T12:42:38Z",
    :createdAt=>"2019-04-03T01:15:52Z"}],
 :success=>true}

=> {:requestId=>"REDACTED", :result=>[{:id=>REDACTED, :firstName=>"REDACTED", :lastName=>"REDACTED", :email=>"REDACTED", :updatedAt=>"2019-06-19T12:42:38Z", :createdAt=>"2019-04-03T01:15:52Z"}], :success=>true}
```

I noticed that get_lead_by_id doesn't display this behavior:

```ruby
pry(main)> response = $marketo.get_lead_by_id(REDACTED, fields: ['phone'])
GET https://REDACTED.mktorest.com/rest/v1/lead/REDACTED.json?fields=phone
User-Agent: "Faraday v0.15.4"
Authorization: "Bearer REDACTED"
Status 200
server: "nginx"
date: "Tue, 16 Jul 2019 23:18:43 GMT"
content-type: "application/json;charset=UTF-8"
content-length: "94"
connection: "close"
{:requestId=>"REDACTED",
 :result=>[{:id=>REDACTED, :phone=>"REDACTED"}],
 :success=>true}

=> {:requestId=>"REDACTED", :result=>[{:id=>REDACTED, :phone=>"REDACTED"}], :success=>true}
```

I noticed that the difference between the 2 requests is that get_leads specifies the fields parameter with `fields%5B%5D=phone` (when decoded this reads fields[]=phone`) while get_lead_by_id specifies the fields parameter with `fields=phone`.

I also gave it a shot to see what would happen if I passed in multiple fields. Calling get_leads with `fields: ['phone', 'title']` sets the url parameter to `fields%5B%5D=phone&fields%5B%5D=title` and doesn't return the fields in the response. Calling get_lead_by_id with `fields: ['phone', 'title']` sets the url parameter to `fields=phone%2Ctitle` (which translates to `fields=phone,title`) and returns the correct fields in the response including the phone and title.

Digging into the documentation at https://developers.marketo.com/rest-api/lead-database/leads/#query I found the following relevant quote that further supports that the comma separated fields parameter is the correct way to query additional fields:
> Optionally you may pass a fields parameter containing a comma-separated list of field names to return.  If the fields parameter is not included in this request, the following default fields will be returned: email, updatedAt, createdAt, lastName, firstName, and id.  When requesting a list of fields, if a particular field is requested, but not returned, the value is implied to be null.

# root cause

I think I found the issue.
The default params encoder built into faraday encodes arrays the way get_leads encodes it: https://github.com/lostisland/faraday/blob/master/lib/faraday/encoders/nested_params_encoder.rb#L65
The reason get_lead_by_id isn't affected is because it converts the array into a string before it gets to the faraday params encoder https://github.com/raszi/mrkt/blob/2e67b842f8b53da9bc03e8ec390a43b62fcb2556/lib/mrkt/concerns/crud_leads.rb#L5

# solutions

I think there are 2 solutions:
1. Update get_leads to convert the fields into a string the same way get_lead_by_id does it with `fields: fields&.join(',')`. This has the problem that we would have to remember to convert any array in the future into a string with a `&.join(',')`.
2. Create a custom marketo param encoder that we set on faraday. This will allow us to remove workarounds like `fields&.join(',')` and solve the problem for all array encoding across the mrkt gem. I found more docs about this at https://github.com/lostisland/faraday/tree/v0.15.4#changing-how-parameters-are-serialized This can be set on the connection options so that it doesn't affect any other faraday usage outside of the mrkt gem.

# workarounds

A simple workaround is to just call `$marketo.get_leads(:email, ['REDACTED'], fields: 'phone,title')`

**Comment by raszi** [comment]
Very descriptive and great issue! Sorry for the late response, I am not using this gem so I check the issues rarely.

I'll take a look on how Marketo is handling these array parameters in general and try to come up with a solution.

**Comment by raszi** [comment]
I believe I came up with a working solution. Could you please check this out?

**Comment by raszi** [comment]
@thomasdziedzic-pd did you have a chance to take a look on the PR? Thank you!

**Comment by thomasdziedzic-pd** [comment]
I tested the code against your branch and it seems like get_leads is now displaying the correct behavior.

Thanks for your time!

**Comment by raszi** [comment]
Thank you for reporting and testing!

**Comment by raszi** [comment]
I've released a new version with this fix included as [v1.0.1](/raszi/mrkt/commit/82babce209eaa54ebbf734ee860be40b6102a185).
